lnrpc/devrpc/dev_server.go

```diff
@@ -0,0 +1,342 @@
+		"gRPC server")
+
+	return nil
+}
+
+// RegisterWithRestServer will be called by the root REST mux to direct a sub
+// RPC server to register itself with the main REST mux server. Until this is
+// called, each sub-server won't be able to have requests routed towards it.
+//
+// NOTE: This is part of the lnrpc.GrpcHandler interface.
+func (r *ServerShell) RegisterWithRestServer(ctx context.Context,
+	mux *runtime.ServeMux, dest string, opts []grpc.DialOption) error {
+
+	// We make sure that we register it with the main REST server to ensure
+	// all our methods are routed properly.
+	err := RegisterDevHandlerFromEndpoint(ctx, mux, dest, opts)
+	if err != nil {
+		log.Errorf("Could not register DEV REST server with the root "+
+			"REST server: %v", err)
+		return err
+	}
+
+	log.Debugf("DEV REST server successfully registered with the root " +
+		"REST server")
+	return nil
+}
+
+// CreateSubServer populates the subserver's dependencies using the passed
+// SubServerConfigDispatcher. This method should fully initialize the
+// sub-server instance, making it ready for action. It returns the macaroon
+// permissions that the sub-server wishes to pass on to the root server for all
+// methods routed towards it.
+//
+// NOTE: This is part of the lnrpc.GrpcHandler interface.
+func (r *ServerShell) CreateSubServer(configRegistry lnrpc.SubServerConfigDispatcher) (
+	lnrpc.SubServer, lnrpc.MacaroonPerms, error) {
+
+	subServer, macPermissions, err := createNewSubServer(configRegistry)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	r.DevServer = subServer
+	return subServer, macPermissions, nil
+}
+
+func parseOutPoint(s string) (*wire.OutPoint, error) {
+	split := strings.Split(s, ":")
+	if len(split) != 2 {
+		return nil, fmt.Errorf("expecting outpoint to be in format of: " +
+			"txid:index")
+	}
+
+	index, err := strconv.ParseInt(split[1], 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode output index: %v", err)
+	}
+
+	txid, err := chainhash.NewHashFromStr(split[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse hex string: %v", err)
+	}
+
+	return &wire.OutPoint{
+		Hash:  *txid,
+		Index: uint32(index),
+	}, nil
+}
+
+func parsePubKey(pubKeyStr string) ([33]byte, error) {
+	var pubKey [33]byte
+	pubKeyBytes, err := hex.DecodeString(pubKeyStr)
+	if err != nil || len(pubKeyBytes) != 33 {
+		return pubKey, fmt.Errorf("invalid pubkey: %v", pubKeyStr)
+	}
+
+	copy(pubKey[:], pubKeyBytes)
+	return pubKey, nil
+}
+
+// ImportGraph imports a graph dump (without auth proofs).
+//
+// NOTE: Part of the DevServer interface.
+func (s *Server) ImportGraph(ctx context.Context,
+	graph *lnrpc.ChannelGraph) (*ImportGraphResponse, error) {
+
+	// Obtain the pointer to the global singleton channel graph.
+	graphDB := s.cfg.GraphDB
+
+	var err error
+	for _, rpcNode := range graph.Nodes {
+		node := &channeldb.LightningNode{
+			HaveNodeAnnouncement: true,
+			LastUpdate: time.Unix(
+				int64(rpcNode.LastUpdate), 0,
```

Comment:
Will we always have a non-zero `rpcNode.LastUpdate`? If not, this will set to 1 sept 1970. Perhaps we should validate since its coming from user input?

Reply:
How would validation look like for these timestamps? I suppose we could use a threshold of say 2019. Jan 1., but that's also not perfect. Also validating the timestamp may restrict testing.

Reply:
Can we just check that the values are non-zero? Or do nodes sometimes not have a last update time?

If they can have zero times, then I think we should only set `LastUpdate = time.Unix(rpcNode.LastUpdate,0)` if its non-zero. That way we have an empty `time.Time{}` and `.IsZero` will work. This way, we set all zero values to unix epoch zero, which doesn't really make sense.